src/backend/validate.ts

```diff
@@ -0,0 +1,60 @@
+import { NegatableOption, RequestSpec } from '../RequestSpec';
+import { Edition } from '../types';
+import { labelForRequestSpecKey } from '../ui/utils';
+
+// Type predicate
+// eslint-disable-next-line
+const isNegatableOption = (value: any): value is NegatableOption => {
+  return typeof value.negated === 'boolean';
+};
+
+const isDefaultValue = (value: string | NegatableOption | string[] | undefined): boolean => {
+  if (typeof value === 'undefined') {
+    return true;
+  }
+  if (typeof value === 'string') {
+    return value === '';
+  }
+  if (isNegatableOption(value)) {
+    return value.value !== undefined || value.value !== '';
+  }
+  if (Array.isArray(value)) {
+    return value.length > 0;
+  } else {
+    throw new Error(`Impossible, Should never happen: isDefaultValue got an Argument of type ${typeof value}`);
+  }
+};
+
+type PotentiallyRequiredKeys = 'graph_type' | 'aggregation' | 'site' | 'service' | 'host_name' | 'graph';
+
+export const getRequiredFields = (edition: Edition) => {
+  const result: PotentiallyRequiredKeys[] = ['graph_type', 'aggregation'];
+  if (edition === 'CEE') {
+    result.push('graph');
+  } else {
+    result.push('site', 'service', 'host_name', 'graph');
+  }
+  return result.sort();
+};
+
+const missingRequiredFields = (rq: Partial<RequestSpec>, edition: Edition): string[] => {
+  const result: Array<keyof RequestSpec> = [];
+  for (const field of getRequiredFields(edition)) {
+    if (rq[field] === undefined || isDefaultValue(rq[field])) {
+      result.push(field);
+    }
+  }
+  return result.map((value) => labelForRequestSpecKey(value, rq));
+};
+
+export const validateRequestSpec = (rq: Partial<RequestSpec> | undefined, edition: Edition): void => {
+  if (rq === undefined) {
+    rq = { graph_type: 'predefined_graph', aggregation: 'off' };
```

Comment:
there is code duplicating this somewhere else, is it? is there a way to remove this here?

Reply:
sorry to bother you again about this: what about completely ignoring `graph_type` and `aggregation` here?
we are pretty sure that those two values have a default value, even if nothing is defined, so we could exclude them from the test? the gui does not allow to undefine those values..

Reply:
Don't worry about it!

I guess you're right. Was maybe thinking about what might happen in the future, but these fields not having a default value seems quite unlikely.